src/wui/savegametable.cc

```diff
@@ -0,0 +1,261 @@
+                                       const SavegameData& savegame) {
+	size_t last_column_index = number_of_columns() - 1;
+	for (size_t col_index = 0; col_index < last_column_index; col_index++) {
+		te.set_string(col_index, "");
+	}
+	te.set_string(
+	   last_column_index, (boost::format(_("Incompatible: %s")) % savegame.mapname).str());
+}
+
+void SavegameTable::create_directory_entry(UI::Table<const uintptr_t>::EntryRecord& te,
+                                           const SavegameData& savegame) {
+	size_t last_column_index = number_of_columns() - 1;
+	for (size_t col_index = 0; col_index < last_column_index; col_index++) {
+		te.set_string(col_index, "");
+	}
+	if (savegame.is_parent_directory()) {
+		te.set_picture(last_column_index, g_gr->images().get("images/ui_basic/ls_dir.png"),
+		               (boost::format("%s") % _("<parent>")).str());
+	} else if (savegame.is_sub_directory()) {
+		te.set_picture(
+		   last_column_index, g_gr->images().get("images/ui_basic/ls_dir.png"),
+		   (boost::format("%s") % FileSystem::filename_without_ext(savegame.filename.c_str())).str());
+	}
+}
+
+void SavegameTable::set_show_filenames(bool) {
+}
+
+SavegameTableSinglePlayer::SavegameTableSinglePlayer(UI::Panel* parent,
+                                                     UI::PanelStyle style,
+                                                     bool localize_autosave)
+   : SavegameTable(parent, style, localize_autosave) {
+
+	add_columns();
+}
+
+void SavegameTableSinglePlayer::add_columns() {
+	add_column(130, _("Save Date"), _("The date this game was saved"), UI::Align::kLeft);
+	add_column(0, _("Description"),
+	           _("The filename that the game was saved under followed by the map’s name, "
+	             "or the map’s name followed by the last objective achieved."),
+	           UI::Align::kLeft, UI::TableColumnType::kFlexible);
+
+	set_sort_column(0);
+}
+
+void SavegameTableSinglePlayer::create_valid_entry(UI::Table<uintptr_t const>::EntryRecord& te,
+                                                   const SavegameData& savegame) {
+	if (savegame.is_directory()) {
+		create_directory_entry(te, savegame);
+	} else {
+		te.set_string(0, savegame.savedatestring);
+		te.set_picture(1, g_gr->images().get("images/ui_basic/ls_wlmap.png"),
+		               map_filename(savegame.filename, savegame.mapname));
+	}
+}
+
+SavegameTableReplay::SavegameTableReplay(UI::Panel* parent,
+                                         UI::PanelStyle style,
+                                         bool localize_autosave)
+   : SavegameTable(parent, style, localize_autosave), show_filenames_(false) {
+	add_columns();
+}
+
+void SavegameTableReplay::add_columns() {
+	add_column(130, _("Save Date"), _("The date this game was saved"), UI::Align::kLeft);
+	std::string game_mode_tooltip = "";
+	/** TRANSLATORS: Tooltip header for the "Mode" column when choosing a game/replay to
+	load.
+	            */
```

Comment:
The comment closing can go in the line above.

Reply:
done